// vue-uploader/webpack.config.js
var path = require("path")
var package = require("./package.json")
var webpack = require("webpack")
module.exports = {
    context: __dirname + "/src",
    entry: "./entry.js",
    output: {
        path: __dirname + "/dist",
        filename: "vue-uploader.js",
        libraryTarget: "umd",
        library: "VueUploader"
    },
    module: {
        loaders: [
            {test: /\.css$/, loader: "style-loader!css-loader"},
            {test: /(\.html)$/, loader: "html-loader"},
            {test: /\.(jpg)|(png)|(gif)$/, loader: "url-loader"},
            {test: /\.vue$/, loader: "vue-loader"},
            {
                test: /\.js$/,
                exclude: /node_modules/,
                loader: 'babel'
            }
        ]
    },
    plugins: [
        new webpack.BannerPlugin("Vue-uploader " + package.version + "\r\n(c) Peak Tai\r\n" + new Date()),
        new webpack.DefinePlugin({
            ROOT: JSON.stringify(path.normalize(__dirname)),
        })
    ],
    debug: false,
    babel: {
        presets: ['es2015'],
        plugins: ['transform-runtime']
    }
}
