// vue-uploader/src/UploadFile.js
/**
 * Created by peak on 16/6/30.
 */
import {sizeConvert} from "./filters";
import {getExtension} from "./util";
import lrz from "./lrz.all.bundle";
class UploadFile {

    /**
     *构建UploadFile
     * @param file {File} 文件对象
     */
    constructor(file) {
        this.name = file.name
        this.type = file.type
        this.size = file.size
        this.hr_size = sizeConvert(file.size)
        this.status = "ready"
        this.progress = 0
        this.extension = getExtension(file.name)
        this.thumbnail = null
        this.origin = file
    }

    /**
     * 创建缩略图,异步进行,完成后回调
     * @param calback {Function}
     */
    makeThumbnail(calback) {
        let uploadFile = this
        if (!uploadFile.type || !uploadFile.type.indexOf("image") == 0) {
            return
        }
        lrz(uploadFile.origin, {height: 50, quality: 90}).then((res)=> {
            uploadFile.thumbnail = res.base64
            if (typeof calback === 'function') {
                calback()
            }
        })
    }

    /**
     * 开始上传
     */
    start() {
        let uf = this
        if (this.status == "finished" || this.status == "uploading") {
            return false
        }

        let xhr = uf.xhr = new XMLHttpRequest()
        xhr.onprogress = function (e) {
            uf.status = "uploading"
            if (e.lengthComputable) {
                uf.progress = Math.floor(e.loaded / e.total);
            }
        }
        xhr.onabort = function (e) {
            uf.status = "abort"
            uf.progress = 0
        }
        xhr.onload = function (e) {
            if (xhr.status != 200) {
                uf.status = "error"
                return
            }
            let json = JSON.parse(xhr.responseText)
            if (!json.ok) {
                uf.status = "error"
                return
            }
            uf.status = "finished"
        }
        xhr.onerror = function (e) {
            uf.status = "error"
        }
        let compress = uf._vm.compress || uf._vm.default_compress
        if (compress && uf.type && uf.type.indexOf("image/") == 0) {
            //compress
            lrz(uf.origin, compress).then(function (rst) {
                sendFile(xhr, rst.file)
                uf._vm.$emit("start", uf)
            }).catch(function (err) {
                console.log(err)
                uf.status = "error"
            })
        } else {
            sendFile(xhr, uf.origin)
            uf._vm.$emit("start", uf)
        }
        return true

        function sendFile(xhr, blob) {
            let formData = new FormData()
            formData.append(uf._vm.fieldName, blob)
            xhr.open("POST", uf._vm.server)
            xhr.send(formData)
        }
    }

    /**
     * 中断上传
     */
    abort() {
        if (this.status != "uploading") {
            return false
        }
        this.xhr.abort()
        return true
    }

    remove() {
        if (this.status == 'uploading') {
            return false
        }
        this._vm.list.$remove(this)
        this._vm.$dispatch("fileRemoved", this)
        return true
    }

    /**
     * 比较
     * @param other
     */
    equals(other) {
        if (!other instanceof UploadFile && !other instanceof File) {
            return false
        }
        return this.name === other.name && this.type === other.type && this.size === other.size
    }


}

export default UploadFile
